util/intern.cc

```diff
@@ -0,0 +1,71 @@
+#include "intern.h"
+#include "xxhash.h"
+#include <cassert>
+#include <cstring>
+#include <mutex>
+#include <unordered_map>
+
+namespace atlas {
+namespace util {
+
+struct Hasher {
+  size_t operator()(const char* str) const {
+    return XXH64(str, strlen(str), 42);
+  }
+};
+
+struct Comparer {
+  bool operator()(const char* s1, const char* s2) const {
+    return std::strcmp(s1, s2) == 0;
+  }
+};
+
+class StringPool {
+ public:
+  StringPool() noexcept {}
+  ~StringPool() noexcept = default;
+  StringPool(const StringPool& pool) = delete;
+  StringPool& operator=(const StringPool& pool) = delete;
+
+  const StrRef& intern(const char* string) {
+    std::lock_guard<std::mutex> guard(mutex);
```

Comment:
If used in a program with a lot of threads, then this will likely cause quite a bit of contention as it needs to be done regularly for tag values. Might want to look at something like a ConcurrentHashMap if there is a good version available for C++.

Reply:
I haven't found a good version yet. The registry is using a similar pattern. Since our main users are the node client, and our system agent, this shouldn't be a problem. But I agree, having a concurrent hash map would be ideal.